### ISSUE TITLE
Support local DynamoDB and update DB documentation

### ISSUE THREAD

**Issue by andy-haynes** [near/near-contract-helper#621]
As of the end of March we've migrated from Postgres to DynamoDB. However the documentation still references Postgres and there is not currently a local development solution for working with DynamoDB.

This should be addressed so that developers can stand up contract helper locally with a backend database with clear documentation around setting this up and running it.

**Comment by mieubrisse** [comment]
Duplicating note from https://github.com/near/near-contract-helper/issues/618 , that Kurtosis users will see an error message after they create their account until we can get a local DynamoDB

**Comment by mieubrisse** [comment]
Hey folks, wanted to check in on this - users are still getting an error when they create new accounts, which makes them think that account creation failed

**Comment by mieubrisse** [comment]
Hey folks, is there any chance of getting this fixed? The local NEAR experience is pretty degraded for folks, with errors coming through on account creation which makes it look like things have failed. It'd be great if there was even just a basic Postgres thing that we could wire in so that the localnet usecase still works!

**Comment by andy-haynes** [comment]
@mieubrisse the referenced PR adds a local DynamoDB instance for local development when running `yarn start` so that development environments utilize it instead of a production DynamoDB environment. The README has also been updated accordingly.

To run local contract helper in a Docker container however, I'd recommend using the official [DynamoDB Docker container](https://hub.docker.com/r/amazon/dynamodb-local/) for a simpler and more complete solution. So long as the contract helper instance is running with the env var `NODE_ENV=development` set, the local DynamoDB instance will be expected at `localhost:7877` (port is hardcoded currently but could be configured from new env var).

**Comment by gbouv** [comment]
Hey @andy-haynes ! I'm looking into this and trying to upgrade the iimage we're using. We're running the contract-helper with `"NODE_ENV": "development"` as of today, so does that mean I don't need to do anything and a local dynamodb should be started automatically?
I'm still seeing weird errors when creating my account, but maybe it's something else.
Otherwise, if I want to use the official dynamodb container, what param should I set in the contract helper to point it to the dynamodb container?

**Comment by gbouv** [comment]
Okay, I got all other container to work almost as expected, so I think this is the missing part here.

From what I'm seeing, everything works fine, except when I create my account, after inputing the word #N of my passphrase, I am redirected to the /recover-seed-phrase page. But when I ignore this and just go to http://127.0.0.1:8334/, everything works as expected. So, the error is closed to what we were seeing before I believe.

Looking at the console, here is what I can see:
```
tmp_fetch_send_json.js:9          POST http://127.0.0.1:8330/account/seedPhraseAdded 400 (Bad Request)
```

And from the request, I get a 400 from `POST http://127.0.0.1:8330/account/seedPhraseAdded` with the following message:
```
CredentialsError: Missing credentials in config, if using AWS_CONFIG_FILE, set AWS_SDK_LOAD_CONFIG=1
```

When I continue to look at the network request after after going to `http://127.0.0.1:8334/`, I see similar failures from `POST http://127.0.0.1:8330/account/recoveryMethods` with message: `CredentialsError: Missing credentials in config, if using AWS_CONFIG_FILE, set AWS_SDK_LOAD_CONFIG=1`

Do you have any idea what's going on here?

**Comment by gbouv** [comment]
Hey @andy-haynes ! Yes I am under the impression dynamodb is not running at all. 

I am not sure I understand how that would work if I spin up the dynamodb container you linked in my docker, since it wouldn't be running in the same as the container contract helper is running. So, trying to hit localhost:7877, that wouldn't work neither.
Maybe it would be easier if the hostname and port of dynamodb was configurable in the service. Would that be possible by any chance?

**Comment by gbouv** [comment]
Ok, so, when I set the above env var I get another error on the `/seedPhraseAdded`:
```
NetworkingError: connect ECONNREFUSED 127.0.0.1:7877
```
So, according to [this page](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html), it seems local dynamo db needs those var to be set anyway with random value to work. 
However, it's now trying to connect to local dynamodb running in the contract-helper container. And even when I spin up a dynamodb container on the side, it still fails because it's running in a different container.

So, I still think being able to configure on which host dynamodb is running would be helpful here @andy-haynes

**Comment by andy-haynes** [comment]
Thanks @gbouv 

> Maybe it would be easier if the hostname and port of dynamodb was configurable in the service. Would that be possible by any chance?

I'm having issues testing this locally but you can give this a shot: https://github.com/near/near-contract-helper/pull/653

Just specify env vars for `LOCAL_DYNAMODB_HOST` and `LOCAL_DYNAMODB_PORT` in the contract helper instance.

>  In the overrideLocalDynamo, we don't seed the schema, so the workflow wtill fails.

Good point, your fix seems fine (race condition notwithstanding 😉) given this for a local dev shim that will eventually go away when a DB is no longer required. Try out the PR I linked to see if that simplifies what you need to get set up and we'll go from there 🙂

**Comment by gbouv** [comment]
Hey @andy-haynes! Your PR is great, I tested it and it works just fine. I set both env vars and I am now able to run DynamoDB on a separate container 👏 👏 

I guess the only missing part is seeding the schema. If adding [this piece of code](https://github.com/near/near-contract-helper/pull/652/files#diff-c08b36ea71edee06502a3996a59354fe1580f7008cbb8366170587921b5b7b3fR21) seems fine to you, I'm happy with it. 

So, I guess the 2 remaining things are:
- [ ] Getting https://github.com/near/near-contract-helper/pull/653 merged
- [ ] Adding [this piece of code](https://github.com/near/near-contract-helper/pull/652/files#diff-c08b36ea71edee06502a3996a59354fe1580f7008cbb8366170587921b5b7b3fR21) or any other way to seed the DynamoDb schema

**Comment by gbouv** [comment]
@andy-haynes any ETA on your side for addressing the 2 remaining points mentioned in my above comments?

**Comment by gbouv** [comment]
@andy-haynes it is impossible (or at least it adds a bunch of complexity) to handle schema initialisation in the DynamoDB container itself because as you mentioned it, schema initialisation can happen only when the DynamoDB container is already running :(
What we can try to do it to run you run your seed command before we run `yarn start` in the contract helper container, but again this seems a bit hacky.

TBH all software I've been working with handle DB seeding themselves. Every time the software starts, it checks the DB, if schema already exists, all good, if not, it create the schema before continuing its boot process.
I am not familiar with the contract helper but I'd be curious to hear more on why this one cannot do this.

**Comment by andy-haynes** [comment]
The logic in `app.js` is basically a development-only change to the DB URL synchronously. Not to mutate data or create tables. It is the responsibility of the local development process starting the DynamoDB instance to initialize the database for use, which is why both the test suite and local development via `yarn start` do exactly that. You are welcome to open a PR if you have a preferred alternative approach that does not involve adding a race condition. It's worth noting however that the entire data layer is slated for deprecation so this is not a priority for the team.

> What we can try to do it to run you run your seed command before we run yarn start in the contract helper container

If you're running `yarn start` already then you should already have a local DockerDB instance...
```json
"start": "node scripts/run-development-dynamo.js & env $(sed 's/ # .*//' .env) supervisor app | bunyan",
```

**Comment by gbouv** [comment]
Ok, well we can go the hacky route of calling `node scripts/create-dynamodb-tables.js` before starting the container then...
Do you have an ETA for getting #653 merged?

**Comment by gbouv** [comment]
Thanks for merging the PR @andy-haynes . Tried it today, and it seems the script does not create the tables in dynamoDB. I filed a new issue to track: https://github.com/near/near-contract-helper/issues/654
